rust/transaction/src/transaction_builder.rs

```diff
@@ -0,0 +1,215 @@
+// Copyright (C) 2019-2021 Aleo Systems Inc.
+// This file is part of the Aleo library.
+
+// The Aleo library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The Aleo library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the Aleo library. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::{Transaction, TransactionError, TransactionKernel};
+use aleo_network::Network;
+
+use snarkvm_algorithms::{
+    merkle_tree::{MerklePath, MerkleTreeDigest},
+    CommitmentScheme,
+    SignatureScheme,
+};
+use snarkvm_dpc::{
+    testnet1::{
+        transaction::Transaction as DPCTransaction,
+        BaseDPCComponents,
+        PrivateProgramInput,
+        PublicParameters,
+        DPC,
+    },
+    DPCComponents,
+    DPCScheme,
+    LedgerScheme,
+};
+
+use once_cell::sync::OnceCell;
+use rand::Rng;
+
+/// A builder struct for the Transaction data type.
+#[derive(Derivative)]
+#[derivative(Default(bound = "N: Network"))]
+pub struct TransactionBuilder<N: Network> {
+    pub(crate) transaction_kernel: OnceCell<TransactionKernel<N>>,
+    pub(crate) old_death_program_proofs: Vec<PrivateProgramInput>,
+    pub(crate) new_birth_program_proofs: Vec<PrivateProgramInput>,
+
+    pub(crate) errors: Vec<TransactionError>,
+}
+
+impl<N: Network> TransactionBuilder<N> {
+    ///
+    /// Returns a new transaction builder.
+    /// To return a transaction and consume the transaction builder, call the `.build()` method.
+    ///
+    pub fn new() -> Self {
+        Self { ..Default::default() }
+    }
+
+    ///
+    /// Returns a new transaction builder and sets field `network: Network`.
+    /// Otherwise, logs a `TransactionError`.
+    ///
+    pub fn transaction_kernel(mut self, transaction_kernel: TransactionKernel<N>) -> Self {
+        let transaction_kernel_string = format!("transaction_kernel: {}", transaction_kernel.to_string());
+        if self.transaction_kernel.set(transaction_kernel).is_err() {
+            self.errors
+                .push(TransactionError::DuplicateArgument(transaction_kernel_string));
+        }
+        self
+    }
+
+    ///
+    /// Returns a new transaction builder with the added vector of old_death_program_proofs
+    /// Otherwise, logs a `TransactionError`.
+    ///
+    pub fn add_old_death_program_proofs(self, old_death_program_proofs: Vec<PrivateProgramInput>) -> Self {
```

Comment:
The `add_*` part of each of the below methods is likely not needed (it is implied)

Reply:
renamed old_death_program_proof -> input_proof
renamed new_birth_program_proof -> output_proof